### PR TITLE
5490: Add test for mathematical operators in time functions

### PR DESCRIPTION
**Pull request by kirmanak**

### CODE REVIEW

File: src/test/java/com/axibase/tsd/api/method/sql/function/date/IsWeekDayTest.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+import static com.axibase.tsd.api.util.Mocks.metric;
+
+public class IsWeekDayTest extends SqlTest {
+    private static final String METRIC_NAME = metric();
+    private static final String ENTITY_NAME = entity();
+
+    @BeforeClass
+    public static void prepareData() throws Exception {
+        final Series series = new Series(ENTITY_NAME, METRIC_NAME)
+                .addSamples(Sample.ofDateInteger("2018-07-23T12:54:00+0300", 20));
```
Comment (raipc) [inline src/test/java/com/axibase/tsd/api/method/sql/function/date/IsWeekDayTest.java]: Please, add a comment with day of week

File: src/test/java/com/axibase/tsd/api/method/sql/function/date/IsWeekDayTest.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+        SeriesMethod.insertSeriesCheck(series);
+    }
+
+    @Issue("5490")
+    @Test(description = "Test support of add operator in the first argument")
+    public void testIsWeekDayFunctionAddOperator() {
+        final String query = String.format("SELECT IS_WEEKDAY(time + 1000*60*60*24, 'RUS') FROM \"%s\"", METRIC_NAME);
```
Comment (raipc) [inline src/test/java/com/axibase/tsd/api/method/sql/function/date/IsWeekDayTest.java]: Please, add a case when `IS_WEEKDAY(time)` is not equal to `IS_WEEKDAY(time + millis)`

Comment (raipc) [reply src/test/java/com/axibase/tsd/api/method/sql/function/date/IsWeekDayTest.java]: There also should be a case when the function returns false

File: src/test/java/com/axibase/tsd/api/method/sql/function/date/IsWeekDayTest.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+                .addSamples(Sample.ofDateInteger("2018-07-22T12:54:00+0300", 20))
+                // Monday (is weekday, next day is weekday)
+                .addSamples(Sample.ofDateInteger("2018-07-23T12:54:00+0300", 20));
+
+        SeriesMethod.insertSeriesCheck(series);
+    }
+
+    @Issue("5490")
+    @Test(description = "Test support of add operator in the first argument")
+    public void testIsWeekDayFunctionAddOperator() {
```
Comment (unrealwork) [inline src/test/java/com/axibase/tsd/api/method/sql/function/date/IsWeekDayTest.java]: * Parameterize the test with following params: `(String timeExpression, boolean isWeekDay)`
* Make the assert message more meaningful: `Failed to calculate {timeExpression} as param of IS_WEEKDAY function`.

File: src/test/java/com/axibase/tsd/api/method/sql/function/date/SqlIsWeekDayTest.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    }
+
+    @Issue("5490")
+    @Test(
+            description = "Test support of mathematical operators in the first argument",
+            dataProvider = "provideTimeExpressionAndResult"
+    )
+    public void testIsWeekDayFunctionAddOperator(final String timeExpression, final boolean[] isWeekDay) {
+        final String query = String.format("SELECT IS_WEEKDAY(%s, 'RUS') FROM \"%s\"", timeExpression, METRIC_NAME);
+        final String[][] expectedRows = new String[isWeekDay.length][1];
```
Comment (unrealwork) [inline src/test/java/com/axibase/tsd/api/method/sql/function/date/SqlIsWeekDayTest.java]:
```java
Arrays.stream(arr)
                .mapToObj(String::valueOf)
                .map(ArrayUtils::toArray)
                .toArray(String[][]::new);
```

**Comment by rodionos** [comment]
@kirmanak - was the `time+number` defect related to `is_workday` function? If not, remove `is_workday` and implement numeric operations with the time column in general.

**Comment by kirmanak** [comment]
@rodionos I am not sure was it related or not, honestly. I tried `date_parse()` function, it worked, but I didn't try any others. Based on the behaviour of `date_parse()` func I suppose the problem was related to `is_workday()`.

**Comment by rodionos** [comment]
My suggestion is to perform manual tests of `time+number` arithmetic in various clauses, with and without functions. Add tests to cover such use cases. 5492 is not related to `is_workday()`.

File: src/test/java/com/axibase/tsd/api/method/sql/function/date/SqlTimeArithmetic.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+                {"MONTH(time + 1000*60*60*24*20)", new String[]{"8", "8", "8"}, "MONTH"},
+        };
+    }
+
+    @Issue("5490")
+    @Test(
+            description = "Test support of mathematical operators in the params",
+            dataProvider = "provideTimeExpression"
+    )
+    public void testSqlFunctionTimeMathOperations(final String timeExpression, final String[] results, final String function) {
```
Comment (unrealwork) [inline src/test/java/com/axibase/tsd/api/method/sql/function/date/SqlTimeArithmetic.java]: Rename `timeExpression` to params, and build the expression by concatenation.

File: src/test/java/com/axibase/tsd/api/method/sql/function/date/SqlTimeArithmetic.java [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+import java.util.Arrays;
+
+import static com.axibase.tsd.api.util.Mocks.entity;
+import static com.axibase.tsd.api.util.Mocks.metric;
+
+public class SqlTimeArithmetic extends SqlTest {
```
Comment (unrealwork) [inline src/test/java/com/axibase/tsd/api/method/sql/function/date/SqlTimeArithmetic.java]: All tests should have `Test` suffix

File: src/test/java/com/axibase/tsd/api/method/sql/function/date/SqlTimeArithmeticTest.java [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+        };
+    }
+
+    @Issue("5490")
+    @Test(
+            description = "Test support of mathematical operators in the params",
+            dataProvider = "provideParams"
+    )
+    public void testSqlFunctionTimeMathOperations(final String expression, final String[] results, final String function) {
+        final String query = "SELECT " + expression + " FROM " + METRIC_NAME;
```
Comment (unrealwork) [inline src/test/java/com/axibase/tsd/api/method/sql/function/date/SqlTimeArithmeticTest.java]: I meant change signature to `testSqlFunctionTimeMathOperations(final String function, final String params, final String[] results)` and `query` template to `SELECT %s(%s) FROM %s`

File: src/test/java/com/axibase/tsd/api/method/sql/function/date/SqlTimeArithmeticTest.java [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+                // Monday (is weekday, next day is weekday)
+                .addSamples(Sample.ofDateInteger("2018-07-23T12:54:00+0300", 20));
+
+        SeriesMethod.insertSeriesCheck(series);
+    }
+
+    @DataProvider
+    public static Object[][] provideParams() {
+        return new Object[][]{
+                {"IS_WEEKDAY(time + 1000*60*60*24*1, 'RUS')", new String[]{"false", "true", "true"}, "IS_WEEKDAY"},
```
Comment (unrealwork) [inline src/test/java/com/axibase/tsd/api/method/sql/function/date/SqlTimeArithmeticTest.java]: The comment wasn't sent in previous review
Represent each test case as call. to method:

```java
private static Object[] testCase(final String function, final String params, String... column) {
        return new Object[] {function, params, toArray(column)};
    }
```

Test cases:

```java
return new Object[][] {
                testCase("IS_WEEKDAY", "time + 1000*60*60*24*1, 'RUS'", "false", "true", "true")
}
```

**Comment by kirmanak** [comment]
@unrealwork @raipc I decided to change `assertMessage` in order to simplify the tests. I guess, `Fail to calculate "IS_WORKDAY(time + 1000*60*60*24*1, 'RUS')"` provides enough information.

File: src/test/java/com/axibase/tsd/api/method/sql/function/date/SqlTimeArithmeticTest.java [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+                .addSamples(Sample.ofDateInteger("2018-07-22T12:54:00+0300", 20))
+                // Monday (is weekday, next day is weekday)
+                .addSamples(Sample.ofDateInteger("2018-07-23T12:54:00+0300", 20));
+
+        SeriesMethod.insertSeriesCheck(series);
+    }
+
+    @DataProvider
+    public static Object[][] provideParams() {
+        return new Object[][]{
```
Comment (unrealwork) [inline src/test/java/com/axibase/tsd/api/method/sql/function/date/SqlTimeArithmeticTest.java]: Return `toArray(cases)`

**Comment by kirmanak** [comment]
@raipc Travis CI check has passed, `return new Object[]{}` has been replaced with `return toArray()` as Igor asked. Please, review.

**Comment by rodionos** [comment]
I ran a quick test. This example works ok:

```sql
SELECT *
  FROM cpu_busy
WHERE time BETWEEN '2017-06-15' AND '2017-06-15'
  LIMIT 10
``` 

Change to `time+100` and observe the error.

The coverage is not sufficient. This test should cover use cases where `time` column can be modified with an arithmetic expression. Add tests for time column arithmetic in all clauses: https://axibase.com/docs/atsd/sql/#syntax

**Comment by kirmanak** [comment]
@rodionos 
in this case you have to use `'`:

```sql
SELECT *
  FROM cpu_busy
WHERE 'time + 100' BETWEEN '2017-06-15' AND '2017-06-15'
  LIMIT 10
```

works well
Without `'` you get `NumberFormatException`, which, I think, is expected. The problem was with `ClassCastException`, when ATSD was trying to cast Double to Long (or vice-versa, I am not sure).

**Comment by rodionos** [comment]
> `'time + 100'`

This makes it a literal string , similar to 'a' > 'b'.

**Comment by kirmanak** [comment]
@rodionos in this case tests should expect an error or the results?

**Comment by rodionos** [comment]
No, why should we include `'a' > 'b'` use case in this test? In this PR we want to test that we can add, subtract, multiple and perform other basic arithmetic with the `time` column, and the result of such operation will be used similar as the time column itself - to apply the date filter to queries time series.

**Comment by kirmanak** [comment]
@rodionos I meant this case:

```sql
SELECT *
  FROM cpu_busy
WHERE time+100 BETWEEN '2017-06-15' AND '2017-06-15'
  LIMIT 10
```

Do we support this syntax or users should use `dateadd()` function?

**Comment by rodionos** [comment]
We should fully support numeric operations with the time column. Time column has a LONG datatype.

File: src/test/java/com/axibase/tsd/api/method/sql/function/date/SqlTimeArithmeticTest.java [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+    @Test(
+            description = "Test support of mathematical operators in select having between clause",
+            dataProvider = "provideSelectBetweenClauses"
+    )
+    public void testSqlSelectHavingBetweenClauseTimeMathOperations(final String operation, final String[] results) {
+        final String query = String.format(
+                "SELECT time FROM \"%s\" GROUP BY PERIOD(1 day, 'UTC') HAVING %s BETWEEN '2018-01-01' AND '2019-01-01'",
+                METRIC_NAME, operation);
+        final String[][] expectedRows = expectedRows(results);
+        final String assertMessage = String.format("Fail to calculate \"%s\"", query);
```
Comment (unrealwork) [inline src/test/java/com/axibase/tsd/api/method/sql/function/date/SqlTimeArithmeticTest.java]: `assertSqlQueryRows` displays Query in any case, be more specific in assert messages, describe test case in assert message.

File: src/test/java/com/axibase/tsd/api/method/sql/function/date/SqlTimeArithmeticTest.java [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+    @Test(
+            description = "Test support of mathematical operators in select having between clause",
+            dataProvider = "provideSelectBetweenClauses"
+    )
+    public void testSqlSelectHavingBetweenClauseTimeMathOperations(final String operation, final String[] results) {
+        final String query = String.format(
+                "SELECT time FROM \"%s\" GROUP BY PERIOD(1 day, 'UTC') HAVING %s BETWEEN '2018-01-01' AND '2019-01-01'",
+                METRIC_NAME, operation);
+        final String[][] expectedRows = expectedRows(results);
+        assertSqlQueryRows(ASSERT_MESSAGE, expectedRows, query);
```
Comment (unrealwork) [inline src/test/java/com/axibase/tsd/api/method/sql/function/date/SqlTimeArithmeticTest.java]: I meant that every assert message should inform what happened and which expression/function e.t.c failed the assertion, in which clause. In the same way that you have done in #506

Comment (kirmanak) [reply src/test/java/com/axibase/tsd/api/method/sql/function/date/SqlTimeArithmeticTest.java]: @unrealwork Should I use `assertOkRequest()` here as I did in #506?

**Pull request merged**